### ISSUE TITLE
Tinydtls repository is gone, install-aiocoap.sh broken

### ISSUE THREAD

**Issue by Decker108** [ggravlingen/pytradfri#44]
The Tinydtls repo (https://git.fslab.de/jkonra2m/tinydtls), referred to in the install-aiocoap.sh script, times out on connect. This breaks the project.

Is there an alternative mirror for Tinydtls?

**Comment by lwis** [comment]
Pydtls is the alternative, and end goal for aiocoap - but they don't yet support Python 3.

**Comment by Decker108** [comment]
Considering this is a Python 3 project, that doesn't make it much of an alternative, does it?

**Comment by lwis** [comment]
There's a view to port the project to Python 3.

**Comment by Decker108** [comment]
So until that happens, aio support for pytradfri is broken?

**Comment by balloob** [comment]
We should find a git clone and put it on github

**Comment by balloob** [comment]
I found a checkout on my computer, probably recent: https://github.com/balloob/tinydtls/tree/cython

Don't have time to check if it works right now so please test. If it works we can update `install-aiocoap.sh` to point at it.

**Comment by ggravlingen** [comment]
Didn't work for me, but it could be that my dev env is messed up.

```
patrik@patrikvm:~/Github-host/tinydtls$ autoreconf
aclocal: warning: autoconf input should be named 'configure.ac', not 'configure.in'
patrik@patrikvm:~/Github-host/tinydtls$ ./configure --with-ecc --without-debug
checking for doxygen... :
checking for etags... /bin/false
checking whether make sets $(MAKE)... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for ranlib... ranlib
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking whether byte ordering is bigendian... no
checking for library containing gethostbyname... none required
checking for library containing socket... none required
checking assert.h usability... yes
checking assert.h presence... yes
checking for assert.h... yes
checking arpa/inet.h usability... yes
checking arpa/inet.h presence... yes
checking for arpa/inet.h... yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking for inttypes.h... (cached) yes
checking netdb.h usability... yes
checking netdb.h presence... yes
checking for netdb.h... yes
checking netinet/in.h usability... yes
checking netinet/in.h presence... yes
checking for netinet/in.h... yes
checking stddef.h usability... yes
checking stddef.h presence... yes
checking for stddef.h... yes
checking for stdint.h... (cached) yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking for strings.h... (cached) yes
checking sys/param.h usability... yes
checking sys/param.h presence... yes
checking for sys/param.h... yes
checking sys/socket.h usability... yes
checking sys/socket.h presence... yes
checking for sys/socket.h... yes
checking for unistd.h... (cached) yes
checking sys/time.h usability... yes
checking sys/time.h presence... yes
checking for sys/time.h... yes
checking time.h usability... yes
checking time.h presence... yes
checking for time.h... yes
checking for sys/types.h... (cached) yes
checking for sys/stat.h... (cached) yes
checking for inline... inline
checking for size_t... yes
checking for struct sockaddr_in6.sin6_len... no
checking for stdlib.h... (cached) yes
checking for GNU libc compatible malloc... yes
checking for memset... yes
checking for select... yes
checking for socket... yes
checking for strdup... yes
checking for strerror... yes
checking for strnlen... yes
checking for fls... no
checking for vprintf... yes
configure: creating ./config.status
.in'ig.status: error: cannot find input file: `Makefile
patrik@patrikvm:~/Github-host/tinydtls$ cd cython
patrik@patrikvm:~/Github-host/tinydtls/cython$ python3 setup.py install
Compiling dtls.pyx because it changed.
[1/1] Cythonizing dtls.pyx
running install
running build
running build_py
creating build
creating build/lib.linux-i686-3.5
copying DTLSSocket.py -> build/lib.linux-i686-3.5
running build_ext
building 'dtls' extension
creating build/temp.linux-i686-3.5
creating build/aes
creating build/sha2
i686-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DDTLSv12=1 -DWITH_SHA256=1 -DDTLS_CHECK_CONTENTTYPE=1 -D_GNU_SOURCE=1 -UNDEBUG -I/usr/include/python3.5m -c dtls.c -o build/temp.linux-i686-3.5/dtls.o
In file included from dtls.c:280:0:
../session.h:22:22: fatal error: tinydtls.h: Filen eller katalogen finns inte
 #include "tinydtls.h"
                      ^
compilation terminated.
error: command 'i686-linux-gnu-gcc' failed with exit status 1
```

**Comment by ggravlingen** [comment]
Also tried this. Sorry about the Swedish when running the script.
```
#!/bin/sh
git clone --depth 1 https://github.com/balloob/tinydtls/tree/cython
cd tinydtls
autoreconf
./configure --with-ecc --without-debug
cd cython
python3 setup.py install

cd ../..
git clone --depth 1 -b tinydtls https://github.com/chrysn/aiocoap/
cd aiocoap
python3 -m pip install .
```

Cloning to "cython"...
remote: Not Found
fatal: repository 'https://github.com/balloob/tinydtls/tree/cython/' not found
./foo.sh: 3: cd: can't cd to tinydtls
autoreconf: 'configure.ac' or 'configure.in' is required
./foo.sh: 5: ./foo.sh: ./configure: not found
./foo.sh: 6: cd: can't cd to cython
python3: can't open file 'setup.py': [Errno 2] No such file or directory
Cloning to "aiocoap"...
remote: Counting objects: 95, done.
remote: Compressing objects: 100% (90/90), done.
remote: Total 95 (delta 8), reused 32 (delta 3), pack-reused 0
Packar upp objekt: 100% (95/95), klart.
Kontrollerar om vi fick alla objekt... klart.
The directory '/home/patrik/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/home/patrik/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.

**Comment by Decker108** [comment]
I will try Ballob's suggestion this evening and get back to you to verify if it's an env problem or not.

The swedish is fine. I'm a native speaker ;)

**Comment by Decker108** [comment]
I noticed that the original Tinydtls repo was up again today, so I used that and manage to install aiocoap. However, it first required installing cython3, so perhaps you should add that to the Installation section of the readme?

I also had to chown /usr/local/lib/python3.5/dist-packages/ to my user to avoid running the install script with sudo, but that might be a problem specific to my environment.

**Comment by Decker108** [comment]
Well, I managed to install all the dependencies, but... I'm always getting time-outs when connecting to the hub.

**Comment by ggravlingen** [comment]
Just a check, are you on firmware version 1.1.0015 on the gateway? I've heard rumours that a rollout of a new version is imminent.

**Comment by Decker108** [comment]
Yes, I'm on 1.1.0015.

**Comment by balloob** [comment]
FYI, my mirror is on the same version as the original url so should have worked. Closing this issue as the original issue is resolved.

If you want to update the installation instructions, feel free to open a PR.

**Comment by alanbacon** [comment]
I managed to install working versions aiocoap and tinydtls from source by following the instructions above (or, at this point, by following the possibly slightly more accurate instructions from  [here](https://github.com/ggravlingen/pytradfri/blob/4d4a682f2f646d41d2a6e4a65858e376f56e24ed/script/install-aiocoap.sh)). But I fell into a gotcha...

I googled my way to this thread because I was getting import errors with aiocoap and tinydtls. Obviously by this point I already had the broken version of aiocoap installed. But after compiling and installing the working version I was still getting import errors. This was because python was still using the broken version of aiocoap by default, I had forgotten to delete it and the new install did not overwrite it. 

So assuming you, dear stranger and fellow githubber, have walked a similar path:  **REMEMBER TO UNINSTALL THE DEFAULT BROKEN VERSION OF AIOCOAP.**
